modules/tinymce/src/core/main/ts/html/ExtractInvalidNode.ts

```diff
@@ -0,0 +1,152 @@
+/**
+ * Copyright (c) Tiny Technologies, Inc. All rights reserved.
+ * Licensed under the LGPL or a commercial license.
+ * For LGPL see License.txt in the project root for license information.
+ * For commercial licenses see https://www.tiny.cloud/
+ */
+
+import { Fun } from '@ephox/katamari';
+
+import AstNode from '../api/html/Node';
+import Schema from '../api/html/Schema';
+import Tools from '../api/util/Tools';
+import { hasOnlyChild, isEmpty } from './ParserUtils';
+
+const removeOrUnwrapInvalidNode = (node: AstNode, schema: Schema, originalNodeParent: AstNode = node.parent) => {
+  if (schema.getSpecialElements()[node.name]) {
+    node.empty().remove();
+  } else {
+    // are the children of `node` valid children of the top level parent?
+    // if not, remove or unwrap them too
+    const children = node.children();
+    for (const childNode of children) {
+      if (!schema.isValidChild(originalNodeParent.name, childNode.name)) {
+        removeOrUnwrapInvalidNode(childNode, schema, originalNodeParent);
+      }
+    }
+    node.unwrap();
+  }
+};
+
+const extractInvalidNode = (node: AstNode, schema: Schema, fixed: Set<AstNode> = new Set(), onCreate: (newNode: AstNode) => void = Fun.noop) => {
+  const textBlockElements = schema.getTextBlockElements();
+  const nonEmptyElements = schema.getNonEmptyElements();
+  const whitespaceElements = schema.getWhiteSpaceElements();
+  const nonSplitableElements = Tools.makeMap('tr,td,th,tbody,thead,tfoot,table');
+
+  let parent: AstNode | undefined, newParent: AstNode | undefined, tempNode: AstNode | undefined;
+
+  // Don't bother if it's detached from the tree
+  if (!node.parent || fixed.has(node)) {
+    return;
+  }
+
+  // If the invalid element is a text block, and the text block is within a parent LI element
+  // Then unwrap the first text block and convert other sibling text blocks to LI elements similar to Word/Open Office
+  if (textBlockElements[node.name] && node.parent.name === 'li') {
+    // Move sibling text blocks after LI element
+    let sibling = node.next;
+    while (sibling) {
+      if (textBlockElements[sibling.name]) {
+        sibling.name = 'li';
+        fixed.add(sibling);
```

Comment:
This is mutating the `fixed` data which isn't ideal... can we get rid of the mutation here and return the result instead? That said we're mutating `node` like crazy here so it's probably not an issue...

Reply:
`fixed` is now entirely internal to the function, so there's _less_ external mutation (nodes still get altered quite a bit, but that's kind of the point)